### PR TITLE
JDK-8313349: Introduce `abstract void HtmlDocletWriter.buildPage()`

### PR DESCRIPTION
**Pull request by jonathan-gibbons**
Please review a medium size update that is the last in the series of recent updates to refactor and simplify  builders and writers.

Having cleared away some of the earlier obfuscation, it becomes clear that all subtypes of `HtmlDocletWriter` generate pages for the overall documentation, and so a new abstract method `buildPage` is introduced to make that more obvious. This resulted in renaming some existing `build` methods and/or removing some static `generate` methods, though some of the more complex ones still remain.

In addition, all writers are now generally created by the `WriterFactory` class, providing an extension point should it ever be needed to subtype these classes. The methods on `WriterFactory` return the appropriate supertype (`HtmlDocletWriter` or `AbstractMemberWriter`) thus enforcing the use of the standard abstract methods that are now available.

There is also some cleanup to some of the member writers, including introducing more abstract methods to define common behavior, simplifying some method signatures, removing unused code, and providing documentation comments.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8313349](https://bugs.openjdk.org/browse/JDK-8313349): Introduce `abstract void HtmlDocletWriter.buildPage()` (**Enhancement** - P4)

### Reviewers
 * [Pavel Rappo](https://openjdk.org/census#prappo) (@pavelrappo - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15114/head:pull/15114` \
`$ git checkout pull/15114`

Update a local copy of the PR: \
`$ git checkout pull/15114` \
`$ git pull https://git.openjdk.org/jdk.git pull/15114/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15114`

View PR using the GUI difftool: \
`$ git pr show -t 15114`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15114.diff">https://git.openjdk.org/jdk/pull/15114.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15114#issuecomment-1662968897)

### CODE REVIEW

**Review by pavelrappo — CHANGES_REQUESTED** [review]

**Comment by pavelrappo** [comment]
That's a good cleanup, Jon. Do please jcheck ("Whitespace error") to allow this PR to move further along the Skara pipeline.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/SummaryListWriter.java [hunk 1]
```diff
@@ -46,6 +46,8 @@
 /**
  * Base class for generating a summary page that lists elements with a common characteristic,
  * such as deprecated elements, preview elements, and so on.
+ *
+ * @param <B> a builder, to build the specific summary information to be written
```
Comment (pavelrappo) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/SummaryListWriter.java:50]: Rewrite the sentence for clarity.

**Review by pavelrappo — CHANGES_REQUESTED** [review]

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/SummaryListWriter.java (hunk 1 above)
Comment (jonathan-gibbons) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/SummaryListWriter.java:50]: Propose the following:
```
/**
 * Base class for generating a summary page that lists elements with a common characteristic,
 * such as deprecated elements, preview elements, and so on.
 * 
 * Note: the use of "Summary" in this context is distinct from the use of "summary" in the
 * context of {@link AbstractMemberWriter#buildSummary(Content)}.
 *
 * @param <B> a builder, to determine the elements to be included in the summary
 */
 ```

**Review by pavelrappo — APPROVED** [review]
Looks good to me.

**Pull request closed without merging**